doc/architectural/background-concepts.md

```diff
@@ -764,16 +764,378 @@ Next, try running minisat on the following formula:
+the basic statements of the program:
+- `C = a<b` maps to C = A&lt;B, for an appropriate formula A&lt;B.
+- `result.1 = b` becomes R1 = B, and `result.2 = a` becomes R2 = A.
+
+To handle the `if` statement, we simply make the execution of each branch
+conditional on C using the &rArr; junctor:
+```C
+  if (C)
+    result.1 = b;
+  else
+    result.2 = a;
+```
+becomes (C &rArr; R1 = B) and ((not C) &rArr; R2 = A), stating that
+the equation for the first assignment holds when C is true, and that for
+the second assignment holds when C is false.
+
+Finally, the &Phi; node is again resolved using the &rArr; junctor: we
+can encode the `return` statement as
+(C &rArr; return = R1) and ((not C) &rArr; return = R2).
+
+At this point, it remains to tie the statements together; we find that we can
+again simply connect them with ''and'', since the statements are always executed
+in sequence. We get:
+> C = a&lt;b and (C &rArr; R1 = B) and (C &rArr; return = R1) and
+> ((not C) &rArr; R2 = A) and ((not C) &rArr; return = R2).
+
+We can extend this approach quite straightforwardly to other constructs, but
+one obvious problem remains: We have not described how to handle loops. This
+turns out to be a substantial issue for theoretical reasons: Programs without
+loop (and without recursive functions) always run for a limited amount of
+execution steps, so we can easily write down formulas that, essentially,
+track their entire execution history. But programs with loops can take
+arbitrarily many steps, or even run into infinite loops, so we cannot describe
+their behavior using a finite propositional formula in the way we have
+done above.
+
+There are multiple approaches to deal with this problem, all with different
+trade-offs. CPROVER chooses bounded model checking as the underlying approach.
+The idea is that we only consider program
+executions that are, in some measure, ''shorter'' than a given bound. This bound
+then implies an upper bound on the size of the required formulas, which makes
+the problem tractable.
+
+To make this concrete, let's go back to the factorial example. We consider
+the function in the following form (in CPROVER, we actually use a `goto`-based
+IR, but the `while`-based version is a bit simpler to understand):
+```C
+unsigned long factorial(unsigned n) {
+  unsigned long fac = 1;
+  unsigned int i = 1;
+  while (i <= n) {
+    fac *= i;
+    i = i+1;
+  }
+  return fac;
+}
+```
+We set the following ''length bound'': The loops in the program are allowed to
+be executed at most three times; we will ignore all other executions. With this
+in mind, we observe that the following two classes of programs behave
+exactly the same:
+```C
+/* Execute this loop at most n+1 times */
+while (e) {
+  body;
+}
+```
+and
+```C
+if (e) {
+  body;
+  /* Execute this loop at most n times */
+  while (e) {
+    body;
+  }
+}
+```
+This transformation is known as ''loop unrolling'': We can always replace a loop
```

Comment:
There is probably a wikipedia page talking about this for compilers, maybe link to it?

Reply:
Done.